crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -228,19 +228,27 @@ where
         command_kind: CommandKind<'_>,
         arg: Argument,
     ) -> Result<V, CommandArgumentError> {
-        if matches!(arg, Argument::GasCoin) && !matches!(command_kind, CommandKind::TransferObjects)
-        {
-            return Err(CommandArgumentError::InvalidGasCoinUsage);
-        }
-        if self.arg_is_borrowed(&arg) {
-            return Err(CommandArgumentError::InvalidUsageOfBorrowedValue);
-        }
+        let is_borrowed = self.arg_is_borrowed(&arg);
         let (input_metadata_opt, val_opt) = self.borrow_mut(arg, UsageKind::ByValue)?;
         let is_copyable = if let Some(val) = val_opt {
             val.is_copyable()
         } else {
-            return Err(CommandArgumentError::InvalidUsageOfTakenValue);
+            return Err(CommandArgumentError::InvalidValueUsage);
         };
+        // If it was taken, we catch this above.
+        // If it was not copyable and was borrowed, error as it creates a dangling reference in
+        // effect.
+        // We allow copyable values to be copied out even if borrowed, as we do not care about
+        // referential transparency at this level.
+        if !is_copyable && is_borrowed {
+            return Err(CommandArgumentError::InvalidValueUsage);
```

Comment:
I condensed these (InvalidUsageOfTakenValue and InvalidUsageOfBorrowedValue) down to a single error to avoid over exposing implementation details in some cases